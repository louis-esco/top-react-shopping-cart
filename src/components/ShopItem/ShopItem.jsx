import styles from "./ShopItem.module.css";
import PropTypes from "prop-types";
import { useState } from "react";

export default function Item({ item, addToCart }) {
  const [qty, setQty] = useState(0);

  const decrement = () => {
    if (qty > 0) setQty(Number(qty) - 1);
  };

  const increment = () => {
    setQty(Number(qty) + 1);
  };

  const handleChange = (e) => {
    setQty(Number(e.target.value));
  };

  const handleClick = () => {
    if (qty > 0) addToCart(item, qty);
  };

  return (
    <div className={styles.item}>
      <div className={styles.itemImg}>
        <img src={item.image} alt="Picture of the item" />
      </div>
      <div className={styles.itemName}>{item.title}</div>
      <div className="itemPrice">{item.price}â‚¬</div>
      <div className={styles.itemShop}>
        <div className="itemQty">
          <button onClick={decrement}>-</button>
          <input
            name="qty"
            className={styles.qtyInput}
            type="number"
            value={qty}
            onChange={handleChange}
          />
          <button onClick={increment}>+</button>
        </div>
        <button onClick={handleClick}>Add to cart</button>
      </div>
    </div>
  );
}

Item.propTypes = {
  item: PropTypes.object,
  addToCart: PropTypes.func,
};
