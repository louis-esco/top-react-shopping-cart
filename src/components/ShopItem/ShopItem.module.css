.item {
  display: flex;
  flex-direction: column;
  gap: 20px;
  font-size: 0.9rem;
  font-weight: 300;
}

.itemImg {
  height: 300px;
  display: flex;
  align-items: center;
  justify-content: center;
}

.itemImg img {
  max-width: 100%;
  max-height: 100%;
}

.itemName {
  height: 2rem;
  text-overflow: ellipsis;
  overflow: hidden;
}

input[type="number"]::-webkit-outer-spin-button,
input[type="number"]::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}
input[type="number"] {
  appearance: textfield;
  -moz-appearance: textfield;
}

.shopItem {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.itemAction {
  display: flex;
  gap: 10px;
}

.itemQty {
  height: 25px;
  display: flex;
}

.qtyInput {
  width: 1.5rem;
  text-align: center;
  z-index: -1;
}

.qtyBtn,
.buyBtn {
  background-color: #5b5b5b;
  color: white;
  border: none;
}

.qtyBtn {
  width: 20px;
}

.qtyBtn.leftBtn {
  border-radius: 3px 0 0 3px;
  margin-right: -3px;
}

.qtyBtn.rightBtn {
  border-radius: 0px 3px 3px 0px;
  margin-left: -3px;
}

.qtyBtn:hover,
.buyBtn:hover {
  cursor: pointer;
}

.buyBtn {
  border-radius: 3px;
}
