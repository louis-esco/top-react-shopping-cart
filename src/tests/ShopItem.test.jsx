import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ShopItem from "../components/ShopItem/ShopItem";

describe("ShopItem component", () => {
  const addToCart = vi.fn();
  const item = {
    id: 1,
    title: "Nice item",
    price: 64,
    description: "You can do lots with this item",
    category: "electronics",
    image: "https://fakestoreapi.com/img/61IBBVJvSDL._AC_SY879_.jpg",
    rating: {
      rate: 3.3,
      count: 203,
    },
  };

  it("Render ShopItem", () => {
    const { container } = render(
      <ShopItem item={item} addToCart={addToCart} />
    );

    expect(container).toMatchSnapshot();
  });

  it("Increment and decrement buttons update input qty", async () => {
    const user = userEvent.setup();

    render(<ShopItem item={item} addToCart={addToCart} />);
    const incrementBtn = screen.getByRole("button", { name: "+" });
    const decrementBtn = screen.getByRole("button", { name: "-" });
    const input = screen.getByRole("spinbutton");

    expect(Number(input.value)).toEqual(0);

    await user.click(incrementBtn);

    expect(Number(input.value)).toEqual(1);

    await user.click(decrementBtn);

    expect(Number(input.value)).toEqual(0);
  });

  it("Can't decrement under 0", async () => {
    const user = userEvent.setup();

    render(<ShopItem item={item} addToCart={addToCart} />);
    const decrementBtn = screen.getByRole("button", { name: "-" });
    const input = screen.getByRole("spinbutton");

    expect(Number(input.value)).toEqual(0);

    await user.click(decrementBtn);

    expect(Number(input.value)).toEqual(0);
  });

  it("Add to cart function not called if qty = 0", async () => {
    const user = userEvent.setup();

    render(<ShopItem item={item} addToCart={addToCart} />);
    const addToCartBtn = screen.getByRole("button", { name: "Add to cart" });

    expect(addToCart).not.toHaveBeenCalled();

    await user.click(addToCartBtn);

    expect(addToCart).not.toHaveBeenCalled();
  });

  it("Add to cart  called if qty > 0", async () => {
    const user = userEvent.setup();

    render(<ShopItem item={item} addToCart={addToCart} />);
    const addToCartBtn = screen.getByRole("button", { name: "Add to cart" });
    const incrementBtn = screen.getByRole("button", { name: "+" });
    await user.click(incrementBtn);
    await user.click(addToCartBtn);

    expect(addToCart).toHaveBeenCalled();
  });
});
