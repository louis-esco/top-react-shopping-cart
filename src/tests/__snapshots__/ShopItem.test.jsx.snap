// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShopItem component > Render ShopItem 1`] = `
<div>
  <div
    class="_item_859760"
  >
    <div
      class="_itemImg_859760"
    >
      <img
        alt="Picture of the item"
        src="https://fakestoreapi.com/img/61IBBVJvSDL._AC_SY879_.jpg"
      />
    </div>
    <div
      class="_itemName_859760"
    >
      Nice item
    </div>
    <div
      class="_shopItem_859760"
    >
      <div
        class="itemPrice"
      >
        64
        €
      </div>
      <div
        class="_itemAction_859760"
      >
        <div
          class="_itemQty_859760"
        >
          <button
            class="_qtyBtn_859760 _leftBtn_859760"
          >
            -
          </button>
          <input
            class="_qtyInput_859760"
            name="qty"
            type="number"
            value="0"
          />
          <button
            class="_qtyBtn_859760 _rightBtn_859760"
          >
            +
          </button>
        </div>
        <button
          class="_buyBtn_859760"
        >
          Add to cart
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ShopItem component > Renders ShopItem 1`] = `
<div>
  <div
    class="_item_859760"
  >
    <div
      class="_itemImg_859760"
    >
      <img
        alt="Picture of the item"
        src="https://fakestoreapi.com/img/61IBBVJvSDL._AC_SY879_.jpg"
      />
    </div>
    <div
      class="_itemName_859760"
    >
      Nice item
    </div>
    <div
      class="_shopItem_859760"
    >
      <div
        class="itemPrice"
      >
        64
        €
      </div>
      <div
        class="_itemAction_859760"
      >
        <div
          class="_itemQty_859760"
        >
          <button
            class="_qtyBtn_859760 _leftBtn_859760"
          >
            -
          </button>
          <input
            class="_qtyInput_859760"
            name="qty"
            type="number"
            value="0"
          />
          <button
            class="_qtyBtn_859760 _rightBtn_859760"
          >
            +
          </button>
        </div>
        <button
          class="_buyBtn_859760"
        >
          Add to cart
        </button>
      </div>
    </div>
  </div>
</div>
`;
